packages/hub/src/utils/createXorbs.ts

```diff
@@ -0,0 +1,121 @@
+/**
+ * Todo: add dedup: we actually need to remember chunks already written, and not add themm to the xorb, and also
+ *  take that into account for file reconstruction
+ * Todo: byte grouping?
+ */
+
+import { XET_CHUNK_HEADER_BYTES, XetChunkCompressionScheme } from "./XetBlob";
+import { compress as lz4_compress } from "../vendor/lz4js";
+
+const TARGET_CHUNK_SIZE = 64 * 1024;
+const XORB_SIZE = 64 * 1024 * 1024;
+const MAX_XORB_CHUNKS = 8 * 1024;
+
+export async function* createXorbs(
+	fileSource: Blob
+): AsyncGenerator<{ xorb: Uint8Array; hash: string }, void, undefined> {
+	const chunkModule = await import("../vendor/xet-chunk/chunker_wasm");
+	await chunkModule.init();
+	const chunker = new chunkModule.Chunker(TARGET_CHUNK_SIZE);
+
+	let xorb = new Uint8Array(XORB_SIZE);
+	const sourceChunks: Array<Uint8Array> = [];
+
+	try {
+		const reader = fileSource.stream().getReader();
+		let xorbOffset = 0;
+		let xorbChunks = Array<{ hash: string; length: number }>();
+
+		const addChunks = function* (chunks: Array<{ hash: string; length: number }>) {
+			for (const chunk of chunks) {
+				let chunkToCopy: Uint8Array;
+				if (chunk.length === sourceChunks[0].length) {
+					chunkToCopy = sourceChunks[0];
+					sourceChunks.shift();
+				} else if (chunk.length < sourceChunks[0].length) {
+					chunkToCopy = sourceChunks[0].subarray(0, chunk.length);
+					sourceChunks[0] = sourceChunks[0].subarray(chunk.length);
+				} else {
+					chunkToCopy = new Uint8Array(chunk.length);
```

Comment:
Could be optimized for less memory allocation btw

Reply:
one thing we could do is have a permanent MAX_CHUNK_SIZE Uint8 array that we reuse?